yarn-project/simulator/src/avm/avm_tree.ts

```diff
@@ -0,0 +1,808 @@
+      }
+      case TreeType.LEAF: {
+        return tree.value;
+      }
+      case TreeType.EMPTY: {
+        return this.zeroHashes[depth - 1];
+      }
+    }
+  }
+
+  /**
+   * Extracts the subtree from a given index and depth
+   * @param index - The index of the node to be extracted
+   * @param depth - The depth of the node to be extracted
+   * @returns The subtree rooted at the index and depth
+   */
+  public getNode(index: bigint, depth: number): Tree {
+    const path = this._derivePathBE(index, depth);
+    const truncatedPath = path.slice(0, depth);
+    truncatedPath.reverse();
+    return this._getNode(truncatedPath, this.tree);
+  }
+
+  /**
+   * This is the recursive helper for getNode
+   */
+  private _getNode(nodePath: number[], tree: Tree): Tree {
+    if (nodePath.length === 0) {
+      return tree;
+    }
+    switch (tree.tag) {
+      case TreeType.BRANCH: {
+        return nodePath.pop() === 0 ? this._getNode(nodePath, tree.leftTree) : this._getNode(nodePath, tree.rightTree);
+      }
+      case TreeType.LEAF:
+      case TreeType.EMPTY: {
+        throw new Error('Node not found');
+      }
+    }
+  }
+
+  private _derivePathLE(index: bigint, depth = this.depth): number[] {
+    return this._derivePathBE(index, depth).reverse();
+  }
+
+  private _derivePathBE(index: bigint, depth = this.depth): number[] {
+    return index
+      .toString(2)
+      .padStart(depth, '0')
+      .split('')
+      .map(x => parseInt(x));
+  }
+
+  /**
+   * This is a recursive function that inserts a leaf into the tree given a path
+   * @param value - The value of the leaf to be inserted
+   * @param insertPath - The path to the leaf, this should be ordered from leaf to root (i.e. LE encoded)
+   * @param depth - The depth of the tree
+   * @param tree - The current tree
+   * @param appendMode - If true we append the relevant zeroHashes to the tree as we traverse
+   */
+  private _insertLeaf(value: Fr, insertPath: number[], depth: number, tree: Tree, appendMode = false): Tree {
+    if (insertPath.length > this.depth || depth > this.depth) {
+      throw new Error('PATH EXCEEDS DEPTH');
+    }
+    if (depth === 0 || insertPath.length === 0) {
+      return Leaf(value);
+    }
+    switch (tree.tag) {
+      case TreeType.BRANCH: {
+        return insertPath.pop() === 0
+          ? Branch(this._insertLeaf(value, insertPath, depth - 1, tree.leftTree, appendMode), tree.rightTree)
+          : Branch(tree.leftTree, this._insertLeaf(value, insertPath, depth - 1, tree.rightTree, appendMode));
+      }
+      case TreeType.LEAF:
+      case TreeType.EMPTY: {
+        const zeroLeaf = appendMode ? Leaf(this.zeroHashes[depth - 1]) : Empty();
+        return insertPath.pop() === 0
+          ? Branch(this._insertLeaf(value, insertPath, depth - 1, zeroLeaf, appendMode), zeroLeaf)
+          : Branch(zeroLeaf, this._insertLeaf(value, insertPath, depth - 1, zeroLeaf, appendMode));
```

Comment:
I'd break these ternary operators up into ifs and assign the results of each `_insertLeaf` to a meaningfully named variables like `updatedLeftTree`, etc